draft-ietf-quic-transport.md

```diff
@@ -3257,6 +3287,49 @@ the Source Connection ID used by the peer during the initial
 handshake, it MUST treat that receipt as a connection error of type
 PROTOCOL_VIOLATION.
 
+## CONNECTION_ID_FINISHED Frame {#frame-connection-id-finished}
+
+An endpoint sends a CONNECTION_ID_FINISHED frame (type=0x1b) to indicate that it
+will no longer use a connection ID that was issued by its peer.  This also
+serves as a request to the peer to send additional connection IDs for future use
+(see {{connection-id}}).  New connection IDs can be delivered via the
+NEW_CONNECTION_ID frame ({{frame-new-connection-id}}).
+
+Retiring a connection ID using the CONNECTION_ID_FINISHED frame invalidates any
+stateless reset tokens associated with that connection ID.
```

Comment:
As an open issue, we should decide how long to allow the old token to be effective.  I would suggest that we use the same 3xRTO timer for that as well.  Then we have 3xRTO everywhere in this.

Reply:
I think it's correct as written, actually, though we could make it more explicit if needed.  If a packet with the old CID was delayed and the recipient drops the CID immediately upon receipt, you'll get a SR that you ought to ignore.  So unless we're *also* adding a 3xRTO requirement on the server to continue accepting the retired CID, the sender needs to immediately stop recognizing the SRT associated with the CID.  You might miss a SR from an old packet if the race lands wrong.  But the packet being sent now has a new, still-valid CID and will presumably elicit a current SR of its own, so that's okay.

Reply:
I would actually disable the reset token asap. The only down side of doing that is to fall back to timer based termination if the connection breaks. But we probably need more text on the behavior of the reset token. Assume that a path is defined by a pair of CID. At any given point, a node will have a "main path" on which it is sending data, and a set of paths in various alternate paths in different stages of validation. Let's look at the following scenarios:

1) Node sends a normal packet, on the main path, and receives back a reset.
2) Node sends a probe with a fresh CID and new addresses and receives back a reset
3) Node sends a probe response with a fresh CID in response to a probe and receives back a reset
4) Node has parked an old path that it does not actively use and receives a reset on that path.

Do we want to treat all these scenarios the same, as in tear down the connection? Or do we want to just accept resets on the path that is actively used?